pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -758,7 +759,18 @@ public void replaceSegment(ImmutableSegment segment, @Nullable ThreadSafeMutable
     }
   }
 
-  protected abstract void removeSegment(IndexSegment segment, MutableRoaringBitmap validDocIds);
+  protected void removeSegment(IndexSegment segment, MutableRoaringBitmap validDocIds) {
+    try (UpsertUtils.PrimaryKeyReader primaryKeyReader = new UpsertUtils.PrimaryKeyReader(segment,
+        _primaryKeyColumns)) {
+      removeSegment(segment, UpsertUtils.getPrimaryKeyIterator(primaryKeyReader, validDocIds));
+    } catch (Exception e) {
+      throw new RuntimeException(
+          String.format("Caught exception while removing segment: %s, table: %s", segment.getSegmentName(),
+              _tableNameWithType), e);
+    }
+  }
+
+  protected abstract void removeSegment(IndexSegment segment, Iterator<PrimaryKey> primaryKeyIterator);
```

Comment:
This change is backward incompatible because sub-classes has to implement this new method. To keep it backward compatible, we can add a default implementation to `throw new UnsupportedOperationException()`, and sub-classes need to implement one of the 2 `removeSegment()`. With a default implementation, no need to change the 2 test

Reply:
Nice idea! Thanks for the suggestion. Updated.